### PR TITLE
mComponent code for the message queue

### PR DESCRIPTION
**Pull request by LiuHaolan**
This PR follows the work from last one:
https://github.com/yiying-zhang/Serverless-LegoOS/pull/1
which set up the syscall/opcode/interface in pComponent.

This PR contains the message queue implementation in mComponent kernel data structure, mainly in three parts:
1, message queue api handler, taking care of the pComponent request and send back what they need. in the form of handle_mq_*_request
2, message queue interface, in the form of mc_mq_send/receive/open/close, taking care of synchronization and calling low-level message queue data structure api. managing the name table(queue name - queue head pointer)
3, low level mq data structure api, in the form of append/pop/free_all, taking care of kernel memory management, ignorant of synchronization.

### CODE REVIEW

File: include/lego/msg_q.h [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+/* 
+ * message queue data structure
+ */
+
+struct mc_msg_queue{
+	char* msg_data;
+	unsigned int msg_size;
+	struct list_head list;
+	/* this array is the pid table 
+ 	struct list_head pid_list; */
```
Comment (yiying-zhang) [inline include/lego/msg_q.h]: what was this pid table for?

File: include/lego/msg_q.h [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+unsigned int append(char* msg_data, unsigned int msg_size, struct list_head* name_nid_dict);
+
+/* 
+ * the msg data passed in should copy a new memory here, msg_data should point to a continous memory
+ */
+unsigned int pop(char* msg_data, int* msg_size, struct list_head* name_nid_dict);
+/*
+ * print name_nid_dict
+ */
+
```
Comment (yiying-zhang) [inline include/lego/msg_q.h]: what's this thing? if no need, just delete it

File: include/memory/file_ops.h [hunk 3]
```diff
@@ -14,6 +14,9 @@
 #include <memory/file_types.h>
 
 extern struct lego_file_operations ramfs_file_ops;
+/* hard code different ramfs operation for different pComponent*/
+extern struct lego_file_operations ramfs_file_ops2;
+
```
Comment (yiying-zhang) [inline include/memory/file_ops.h]: what's this for? why is there ops and ops2....

File: managers/memory/file_ops.c [hunk 4]
```diff
@@ -49,11 +49,15 @@ struct lego_file *file_open(struct lego_task_struct *tsk, const char *filename)
 	strncpy(file->filename, filename, MAX_FILENAME_LEN);
 
 #ifdef CONFIG_USE_RAMFS
-	file->f_op = &ramfs_file_ops;
+	unsigned int node0 = 0;
+	if(tsk->node==node0)
+		file->f_op = &ramfs_file_ops;
+	else
+		file->f_op = &ramfs_file_ops2;
```
Comment (yiying-zhang) [inline managers/memory/file_ops.c]: this needs to be changed. btw, should msg be a file data structure? this is Linux way of treating everything as a file. We don't have to do that and can/should create a separate data structure for messages.

File: managers/memory/handle_mq.c [hunk 5]
```diff
@@ -0,0 +1,301 @@
...
+#include <lego/kernel.h>
+#include <memory/thread_pool.h>
+
+unsigned int append(char* msg_data, unsigned int msg_size, struct list_head* name_nid_dict){
+	struct mc_msg_queue* tmp;
+
+	tmp = kmalloc(sizeof(struct mc_msg_queue), GFP_KERNEL);
+	if(!tmp){
+		printk("allocated wrong\n");
+		return MSG_RET_FAIL;
```
Comment (yiying-zhang) [inline managers/memory/handle_mq.c:18]: more informative error msg like "handle_mq append can't allocate memory"

File: managers/memory/handle_mq.c [hunk 6]
```diff
@@ -0,0 +1,301 @@
...
+#include <lego/slab.h>
+#include <lego/uaccess.h>
+#include <lego/files.h>
+#include <lego/syscalls.h>
+#include <lego/comp_common.h>
+#include <lego/kernel.h>
+#include <memory/thread_pool.h>
+
+unsigned int append(char* msg_data, unsigned int msg_size, struct list_head* name_nid_dict){
+	struct mc_msg_queue* tmp;
```
Comment (yiying-zhang) [inline managers/memory/handle_mq.c]: don't use tmp, use something like msg

File: managers/memory/handle_mq.c [hunk 7]
```diff
@@ -0,0 +1,301 @@
...
+}
+
+
+
+/* the msg data passed in should copy a new memory here, msg_data should point to a continous memory
+ */
+unsigned int pop(char* msg_data, int* msg_size, struct list_head* name_nid_dict){
+	struct list_head* next = name_nid_dict->next;
+
+	struct mc_msg_queue* item = list_entry(next, struct mc_msg_queue, list);
```
Comment (yiying-zhang) [inline managers/memory/handle_mq.c:48]: don't you need to check for list empty here?

File: managers/memory/handle_mq.c [hunk 8]
```diff
@@ -0,0 +1,301 @@
...
+
+	strcpy(msg_data, item->msg_data);
```
Comment (yiying-zhang) [inline managers/memory/handle_mq.c:55]: check msg_data and item->msg_data for NULL, just in case

File: managers/memory/handle_mq.c [hunk 9]
```diff
@@ -0,0 +1,301 @@
...
+	*msg_size = item->msg_size;
+
+	list_del(&item->list);
+
+	kfree(item->msg_data);	
+	kfree(item);
+	name_nid_dict = next;
```
Comment (yiying-zhang) [inline managers/memory/handle_mq.c:62]: this seems to be problematic. what if there's another thread who's doing dequeue? I think we now support (or potentially want to support) multiple mComponent handler threads. Need some thread synchronization here.

File: managers/memory/handle_mq.c [hunk 10]
```diff
@@ -0,0 +1,301 @@
...
+/*
+ * print api, for debug
+ */ 
+void print(struct list_head* name_nid_dict){
+	struct mc_msg_queue *pos = NULL;
+	list_for_each_entry(pos, name_nid_dict, list){
+		printk(pos->msg_data);
+		printk(" ");
+		printk("message data: %d\n", pos->msg_size);
+	}
```
Comment (yiying-zhang) [inline managers/memory/handle_mq.c]: need list thread synchronization like other functions

File: managers/memory/handle_mq.c [hunk 11]
```diff
@@ -0,0 +1,301 @@
...
+	while(cur != name_nid_dict){		
+		struct mc_msg_queue* item = list_entry(cur, struct mc_msg_queue, list);
+
+		struct list_head* tmp = cur->next;
+		list_del(cur);
+
+		cur = tmp;
+		kfree(item->msg_data);
+		kfree(item);
+	}
```
Comment (yiying-zhang) [inline managers/memory/handle_mq.c:85]: synchronization

File: managers/memory/handle_mq.c [hunk 12]
```diff
@@ -0,0 +1,301 @@
...
+
+	struct list_head* cur=name_nid_dict->next;
+	while(cur != name_nid_dict){		
+		struct mc_msg_queue* item = list_entry(cur, struct mc_msg_queue, list);
+
+		struct list_head* tmp = cur->next;
+		list_del(cur);
+
+		cur = tmp;
+		kfree(item->msg_data);
```
Comment (yiying-zhang) [inline managers/memory/handle_mq.c:83]: check nULL

File: managers/memory/handle_mq.c [hunk 13]
```diff
@@ -0,0 +1,301 @@
...
+
+/*
+ * api for the message queue data structure in mComponent kernel side
+ */ 
+unsigned int mc_mq_open(char* mq_name, unsigned int max_size)
+{
+
+/*
+ * what if we already got a message queue with that name in the name map?	
+ */
```
Comment (yiying-zhang) [inline managers/memory/handle_mq.c:96]: You should return an error code like MQ_OPEN_ERROR_ALREADY_EXIST

File: managers/memory/handle_mq.c [hunk 14]
```diff
@@ -0,0 +1,301 @@
...
+
+	tmp = kmalloc(sizeof(struct name_mq_map), GFP_KERNEL);
+	if(!tmp){
+		printk("allocated wrong\n");
+		return MSG_RET_FAIL;
+	}
+
+	strcpy(tmp->mq_name, mq_name);
+	tmp->max_size = max_size;
+	tmp->mq = kmalloc(sizeof(struct list_head), GFP_KERNEL);
```
Comment (yiying-zhang) [inline managers/memory/handle_mq.c:125]: check null

File: managers/memory/handle_mq.c [hunk 15]
```diff
@@ -0,0 +1,301 @@
...
+	struct name_mq_map *pos, *target = NULL;
+	list_for_each_entry(pos, &addr_map, list){
+		if(strcmp(mq_name, pos->mq_name)==0){
+			target = pos;		
+		}
+	}
+	if(target != NULL){
+		/* spin lock acquire here */
+		unsigned long flags;
+
```
Comment (yiying-zhang) [inline managers/memory/handle_mq.c:199]: what if the mq get closed after your list search and before your append?

File: managers/memory/handle_mq.c [hunk 16]
```diff
@@ -0,0 +1,301 @@
...
+	struct name_mq_map *pos, *target = NULL;
+	list_for_each_entry(pos, &addr_map, list){
+		if(strcmp(mq_name, pos->mq_name)==0){
+			target = pos;		
+		}
+	}
+	if(target ==NULL){
+		return MSG_RET_FAIL;
+	}
+
```
Comment (yiying-zhang) [inline managers/memory/handle_mq.c:231]: similar to send, potential synchronization problem

File: managers/memory/ramfs/core2.c [hunk 17]
```diff
@@ -0,0 +1,78 @@
...
+{
+	vma->vm_ops = &ramfs_vma_ops2;
+	return 0;
+}
+
+struct lego_file_operations ramfs_file_ops2 = {
+	.read	= ramfs_read2,
+	.write	= ramfs_write2,
+	.mmap	= ramfs_mmap2,
+};
```
Comment (yiying-zhang) [inline managers/memory/ramfs/core2.c]: this is very hacky. What exactly is this file for?

File: include/lego/msg_q.h (hunk 1 above)
Comment (LiuHaolan) [reply include/lego/msg_q.h]: I will delete it

File: managers/memory/handle_mq.c (hunk 9 above)
Comment (LiuHaolan) [reply managers/memory/handle_mq.c:62]: pop function has no synchronization, I put it in the mc_mq_receive

Comment (LiuHaolan) [reply managers/memory/handle_mq.c:62]: every message queues have its own lock, implementing it in mc_mq_receive just to reduce argument passing or traversing in pop.

File: managers/memory/handle_mq.c (hunk 10 above)
Comment (LiuHaolan) [reply managers/memory/handle_mq.c]: this is just for debug. A simple print function. I can delete this one.

File: managers/memory/handle_mq.c (hunk 11 above)
Comment (LiuHaolan) [reply managers/memory/handle_mq.c:85]: right, I add the synchronization

File: managers/memory/handle_mq.c (hunk 15 above)
Comment (LiuHaolan) [reply managers/memory/handle_mq.c:199]: cool, didn't expect that, I add the synchronization

**Pull request closed without merging**